### ISSUE TITLE
Allow states to contain data with lifetimes just like events can

### ISSUE THREAD

**Issue by dzimmanck** [korken89/smlang-rs#26]
While events can contain data with lifetimes, states cannot and generate a "use of undeclared lifetime" error inside the procedural macro if you try.

Simple example:

```rust
//! An example of generating events with actions

#![deny(missing_docs)]

use smlang::statemachine;

/// State data
#[derive(PartialEq)]
pub struct Data<'a>(Option<&'a Events>);

statemachine! {
    transitions: {
        *D0 + ToD1 / to_d1  = D1,
        D1(Data<'a>) + ToD0 = D0,
    }
}

/// Context
pub struct Context;

impl StateMachineContext for Context {
    fn to_d1(&mut self) -> Data {
        Data(Some(&Events::ToD0))
    }
}
```

![image](https://user-images.githubusercontent.com/16160230/147997272-2d3fe6be-7317-4590-8aea-bd7e7e4cd8e4.png)

**Comment by dzimmanck** [comment]
I am working on this.  It involves a pretty substantial re-factor of the parser code in order to re-use the data type parsing for both states and events.

**Comment by dzimmanck** [comment]
I have a branch ready for review at https://github.com/dzimmanck/smlang-rs/tree/feature/support_lifetimes_in_state_data.

I haven't issued a pull request yet because:

1.  I have not verified that it works, only that it does not break existing tests
2.  It needs testing
3.  It needs at least 1 example
4.  It needs documentation

This involved quite a bit of re-factor that I think went pretty well.  To aide in review, I broke it into 8 concise commits with limited scope for each change.  @korken89, would love a review of these changes before I proceed when you get a chance.
